mongo/errors.go

```diff
@@ -71,6 +73,60 @@ func replaceErrors(err error) error {
 	return err
 }
 
+// IsDuplicateKeyError returns true if err is a duplicate key error
+func IsDuplicateKeyError(err error) bool {
+	// handles SERVER-7164 and SERVER-11493
+	for ; err != nil; err = unwrap(err) {
+		if e, ok := err.(ServerError); ok {
+			return e.HasErrorCode(11000) || e.HasErrorCode(11001) || e.HasErrorCode(12582) ||
+				e.HasErrorCodeWithMessage(16460, " E11000 ")
+		}
+	}
+	return false
+}
+
+// IsTimeout returns true if err is from a timeout
+func IsTimeout(err error) bool {
+	for ; err != nil; err = unwrap(err) {
+		// check unwrappable errors together
+		if err == context.DeadlineExceeded {
+			return true
+		}
+		if ne, ok := err.(net.Error); ok {
+			return ne.Timeout()
+		}
+		//timeout error labels
+		if se, ok := err.(ServerError); ok {
+			if se.HasErrorLabel("NetworkTimeoutError") || se.HasErrorLabel("ExceededTimeLimitError") {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// unwrap returns the inner error if err implements Unwrap(), otherwise it returns nil.
+func unwrap(err error) error {
+	u, ok := err.(interface {
+		Unwrap() error
+	})
+	if !ok {
+		return nil
+	}
+	return u.Unwrap()
+}
+
+// IsNetworkError returns true if err is a network error
+func IsNetworkError(err error) bool {
```

Comment:
I'm not sure if this will cover complex cases like network errors during connection handshakes. We should have tests for such cases.

Reply:
added the helper to existing tests. i can also add separate tests for it if we want more